More consistent UserExceptions

magicDGS opened the pull request:
When I was trying to use user exceptions in a consistent way independently of the constructor (mostly related with files), I found very weird behaviour with the messages. Here I try to fix some of the things that I was struggling with:

* Support for path in constructors for `CouldNotReadInputFile`, `CouldNotCreateOutputFile`, `MalformedFile` and `MalformedBAM`, in addition to some missing constructors to have the same structure for all of them (with `File` and/or `String`).
* ~~Updated javadoc in `CommandLineException`, including extending classes to make clear that in the GATK framework is not printed out if it is thrown out of parameter validation.~~ __Edited__: this is not longer required, because `CommandLineException` is decoupled from `UserException` through barclay.
* Added a TODO into the `MalformedBAM` constructor that includes a `GATKRead` that is not used.
* __Edited__: added final to constructors.

cmnbroad commented:
@magicDGS Can you resolve the conflicts in this and then I can review ? (Some of it may need to move to Barclay). thanks.

magicDGS commented:
I resolved the conflicts, @cmnbroad. The only changes for Barclay were javadoc improvements related to the usage of `CommandLineException` after parsing, because I had some issues thinking that any `UserException` will be catch and exit with an error message, but now that it is decoupled into its own class I don't think that it is necessary.

magicDGS commented:
This is a very simple patch, @cmnbroad. Could you have a look?

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -59,22 +60,42 @@ public CouldNotReadInputFile(File file) {
             super(String.format("Couldn't read file %s", file.getAbsolutePath()));
         }
 
+        public CouldNotReadInputFile(Path path) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -59,22 +60,42 @@ public CouldNotReadInputFile(File file) {
+            this(file.getAbsolutePath(), message, e);
+        }
+
+        public CouldNotReadInputFile(Path path, String message, Exception e) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -59,22 +60,42 @@ public CouldNotReadInputFile(File file) {
+            this(path.toString(), message, e);
+        }
+
+        public CouldNotReadInputFile(String file, String message, Exception e) {
```
final, and "file" -> "fileName"

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -59,22 +60,42 @@ public CouldNotReadInputFile(File file) {
             this(file, getMessage(e));
         }
 
+        public CouldNotReadInputFile(Path path, Exception e) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -105,21 +126,37 @@ public CannotHandleGzippedRef() {
 
         public CouldNotCreateOutputFile(File file, String message, Exception e) {
-            super(String.format("Couldn't write file %s because %s with exception %s", file.getAbsolutePath(), message, getMessage(e)));
+            this(file.getAbsolutePath(), message, e);
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -105,21 +126,37 @@ public CannotHandleGzippedRef() {
 
-        public CouldNotCreateOutputFile(File file, String message) {
-            super(String.format("Couldn't write file %s because %s", file.getAbsolutePath(), message));
+        public CouldNotCreateOutputFile(Path path, String message, Exception e) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -59,22 +60,42 @@ public CouldNotReadInputFile(File file) {
             super(String.format("Couldn't read file %s. Error was: %s", file, message));
         }
 
         public CouldNotReadInputFile(File file, String message, Exception e) {
```
final args

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -105,21 +126,37 @@ public CannotHandleGzippedRef() {
             super(String.format("Couldn't write file %s because %s with exception %s", filename, message, getMessage(e)));
         }
 
+        public CouldNotCreateOutputFile(File file, String message) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -105,21 +126,37 @@ public CannotHandleGzippedRef() {
+            this(file.getAbsolutePath(), message);
+        }
+
+        public CouldNotCreateOutputFile(Path path, String message) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -105,21 +126,37 @@ public CannotHandleGzippedRef() {
+            this(path.toString(), message);
+        }
+
+        public CouldNotCreateOutputFile(String file, String message) {
```
final, and "file" ->"fileName"

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -105,21 +126,37 @@ public CannotHandleGzippedRef() {
             super(String.format("Couldn't write file %s because exception %s", file.getAbsolutePath(), getMessage(e)));
         }
 
+        public CouldNotCreateOutputFile(Path path, Exception e) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -187,12 +224,28 @@ public MalformedFile(String message) {
 
-        public MalformedFile(File f, String message) {
-            super(String.format("File %s is malformed: %s", f.getAbsolutePath(), message));
+        public MalformedFile(String file, String message) {
```
final, and "file" -> "fileName"

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -187,12 +224,28 @@ public MalformedFile(String message) {
+            super(String.format("File %s is malformed: %s", file, message));
+        }
+
+        public MalformedFile(String file, String message, Exception e) {
```
final, and "file" -> "fileName"

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -187,12 +224,28 @@ public MalformedFile(String message) {
 
-        public MalformedFile(File f, String message, Exception e) {
-            super(String.format("File %s is malformed: %s caused by %s", f.getAbsolutePath(), message, getMessage(e)));
+        public MalformedFile(File file, String message) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -187,12 +224,28 @@ public MalformedFile(String message) {
+            this(file.getAbsolutePath(), message);
+        }
+
+        public MalformedFile(File file, String message, Exception e) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -187,12 +224,28 @@ public MalformedFile(String message) {
+            this(file.getAbsolutePath(), message, e);
+        }
+
+        public MalformedFile(Path path, String message) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -187,12 +224,28 @@ public MalformedFile(String message) {
+            this(path.toString(), message);
+        }
+
+        public MalformedFile(Path path, String message, Exception e) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -244,12 +297,17 @@ public CannotExecuteRScript(String message) {
             this("(unknown)", message);
         }
 
         public MalformedBAM(File file, String message) {
```
final

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -244,12 +297,17 @@ public CannotExecuteRScript(String message) {
+            this(file.getAbsolutePath(), message);
+        }
+
+        public MalformedBAM(Path path, String message) {
```
final

magicDGS commented:
I addressed your comments, @cmnbroad. In addition, to make more consistent the errors for malformed `GATKRead`, I updated `ReadMissingReadGroup` and removed `MalformedRead` in two separated commits. Like that, every offending read is passing by `MalformedBAM` for consisten error messages.

Back to you and many thanks!

cmnbroad commented:
@magicDGS It looks like you've added a few more commits with additional changes over what I reviewed, and the response to my code review comments folded in amongst them(?). It would be easier to review this if there were one incremental commit (over the ones I reviewed) with the just the changes based on the code review. Thx.

magicDGS commented:
@cmnbroad, in principle https://github.com/broadinstitute/gatk/pull/2282/commits/6d1cbf57e5ba014239cfd0aa5ab9f8280a194032 have just your comments and (although I added final to every param) and other commits have the other changes. I guess that that's what you are referring when you said "incremental commits". If not, just let me know what do you need. Sorry for the inconvenience.

magicDGS commented:
Some advance on thi @cmnbroad?

cmnbroad commented:
@magicDGS Yes, it was the additional commits beyond 6d1cbf5 that I was referring to (and it wasn't clear to me whether **all** of the code review requests were in that commit, or if some were distributed amongst the other commits with the additional changes). In general I think once we start a review, the only changes should be based the code review requests. Thanks!

magicDGS commented:
I removed the latest commits, @cmnbroad. All your comments should be addressed. I will add some questions regarding the commits that I removed.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -104,51 +125,69 @@ public CannotHandleGzippedRef() {
     }
 
+    // todo - why we need this?
+    // todo - remove to use MalformedBAM?
```
This `UserException` is redundant with `MalformedBAM`. I have in my backup branch a commit considering this: https://github.com/magicDGS/gatk/commit/15d7ca3ba732f0f8c3475c637d2655f045787b9b

magicDGS replied:
I think that it is worthy to clean up this here. What do you think, @cmnbroad?

magicDGS commented:
Back to you @cmnbroad, with the two latest commits removed and a couple of questions regarding the other commits (but without including them yet).

magicDGS commented:
Should I rebase because of #2336, @cmnbroad? In addition, I think that I should also look at the propagation of passed exceptions in this PR, because there are some parts of the code that are not really doing it...

cmnbroad commented:
@magicDGS Yes, please rebase, and also do the propagation while we're at it.

magicDGS commented:
Rebased @cmnbroad. The propagation of Exceptions should be fine, I did it in previous commits.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -51,39 +52,59 @@ protected static String getMessage(final Throwable t) {
+        }
+
+        public CouldNotReadInputFile(final File file, final Exception e) {
+            this(file, getMessage(e));
```
Self-note: this requires propagation.

magicDGS replied:
Done

magicDGS commented on src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java:
```diff
@@ -51,39 +52,59 @@ protected static String getMessage(final Throwable t) {
+        }
+
+        public CouldNotReadInputFile(final Path path, final Exception e) {
+            this(path, getMessage(e));
```
Self-note: this requires propagation.

magicDGS replied:
Done

magicDGS commented:
Friendly ping @cmnbroad.

cmnbroad commented:
Hi @magicDGS. After looking a bit more at this PR , and talking with others, I think we should put this branch on hold for a bit. We plan to introduce a new class soon that will serve as a common currency for input specifiers, and we'll probably want to revisit these constructors then. Hopefully that will be a simplifying change. In the meantime, I'd prefer not to introduce more overloads. We'd like to put this on hold until we make progress on that, and then perhaps resurrect parts of this branch as needed. My apologies for the churn.

magicDGS commented:
Thanks for the feedback @cmnbroad. I would like to have something more consistent for my own tools, independently on if I'm using `Path` or `File, but it could wait.

magicDGS commented:
Any progress on this, @cmnbroad?

magicDGS commented:
Thanks for the update, @cmnbroad. I think that at least til that is solved, we could add a temporary solution to be able to use a consistent way of dealing with `File` and `java.nio.Path` user exceptions...

What if I submit a PR to change all ctors to accept `java.nio.Path` instead of `File` in file-related `UserException`s? The `File.toPath()` will always work, but the `Path.toFile()` doesn't. I found problems with respect to HDFS paths when throwing exceptions, and this will allow to use them with `File` and `Path`...

droazen commented:
Closing this ancient PR -- we'll revisit the exception constructors when we add the URI class.
